### PR TITLE
stable formatting for negate()

### LINKED ISSUE #22116: The query formatting is broken
https://clickhouse-test-reports.s3.yandex.net/22102/4b6b1311ce630513139a53a05c4a7be1036b686b/fuzzer_asan/report.html#fail1
```
The query formatting is broken.
Changed settings: max_block_size = '1000000', min_insert_block_size_rows = '0', min_insert_block_size_bytes = '0', max_distributed_connections = '1', connections_with_failover_max_tries = '0', optimize_move_to_prewhere = '0', min_count_to_compile_expression = '1', optimize_distributed_group_by_sharding_key = '1', optimize_skip_unused_shards = '1', log_queries = '1', insert_quorum = '2', insert_quorum_parallel = '0', select_sequential_consistency = '1', joined_subquery_requires_alias = '0', query_profiler_real_time_period_ns = '1000000000', max_rows_to_read = '1', send_logs_level = 'fatal', prefer_localhost_replica = '0', allow_experimental_bigint_types = '1'
Got the following (different) text after formatting the fuzzed query and parsing it back:
'WITH NULL AS a SELECT toDate(-7 IN (inf - NULL, toDate([NULL, '10485.76', '25.5', NULL], NULL, NULL), '65537'), toDate([NULL, '0.0000000255', NULL], '922337203685477580.7', NULL, NULL))'
, expected:
'WITH NULL AS a SELECT toDate((- 7) IN (inf - NULL, toDate([NULL, '10485.76', '25.5', NULL], NULL, NULL), '65537'), toDate([NULL, '0.0000000255', NULL], '922337203685477580.7', NULL, NULL))'
In more detail:
AST-1:
'SelectWithUnionQuery, 0x000060f00002e4f8
-ExpressionList, 0x000060700019e458
--SelectQuery, 0x000060c000173f98
---ExpressionList, 0x000060700052d1e8
----Literal_NULL, 0x00006110008265a0
---ExpressionList, 0x00006070006e6f78
----Function_toDate, 0x0000611000c04a98
-----ExpressionList, 0x00006070000256b8
------Function_in, 0x0000611000466bd8
-------ExpressionList, 0x00006070002c1748
--------Function_negate, 0x0000611000aca998
---------ExpressionList, 0x000060700013cbc8
----------Literal_Int64_7, 0x0000611000aa56a0
--------Function_tuple, 0x0000611000689ed8
---------ExpressionList, 0x00006070005c5b28
----------Function_minus, 0x0000611000a1ced8
-----------ExpressionList, 0x00006070000c9048
------------Literal_Float64_inf, 0x0000611000338660
------------Literal_NULL, 0x0000611000441520
----------Function_toDate, 0x0000611000b08b58
-----------ExpressionList, 0x000060700058d6c8
------------Function_array, 0x0000611000ab9f58
-------------ExpressionList, 0x000060700057e158
--------------Literal_NULL, 0x0000611000b1e0a0
--------------Literal_Decimal64_'10485.76', 0x0000611000386220
--------------Literal_Decimal64_'25.5', 0x00006110005cb3a0
--------------Literal_NULL, 0x00006110003bb920
------------Literal_NULL, 0x00006110009fce60
------------Literal_NULL, 0x0000611000b0df20
----------Literal_Decimal64_'65537', 0x00006110004b7720
------Function_toDate, 0x0000611000ac8f58
-------ExpressionList, 0x00006070006010f8
--------Function_array, 0x0000611000b9a918
---------ExpressionList, 0x000060700041fc18
----------Literal_NULL, 0x00006110006bed20
----------Literal_Decimal64_'0.0000000255', 0x00006110003ad1e0
----------Literal_NULL, 0x0000611000384420
--------Literal_Decimal64_'922337203685477580.7', 0x00006110005b9560
--------Literal_NULL, 0x000061100048afe0
--------Literal_NULL, 0x00006110008cd620
'
Text-1 (AST-1 formatted):
'WITH NULL AS a SELECT toDate((- 7) IN (inf - NULL, toDate([NULL, '10485.76', '25.5', NULL], NULL, NULL), '65537'), toDate([NULL, '0.0000000255', NULL], '922337203685477580.7', NULL, NULL))'
AST-2 (Text-1 parsed):
'SelectWithUnionQuery, 0x000060f0000a3258
-ExpressionList, 0x0000607000571ea8
--SelectQuery, 0x000060c00008ec18
---ExpressionList, 0x00006070000dd328
----Literal_NULL, 0x00006110003c6be0
---ExpressionList, 0x00006070001912c8
----Function_toDate, 0x000061100068fb58
-----ExpressionList, 0x000060700008cdc8
------Function_in, 0x000061100095d5d8
-------ExpressionList, 0x0000607000149818
--------Literal_Int64_-7, 0x0000611000bad020
--------Function_tuple, 0x0000611000707158
---------ExpressionList, 0x0000607000256138
----------Function_minus, 0x000061100032b818
-----------ExpressionList, 0x0000607000347cb8
------------Literal_Float64_inf, 0x00006110007293a0
------------Literal_NULL, 0x0000611000b4c720
----------Function_toDate, 0x000061100001f098
-----------ExpressionList, 0x00006070001ac618
------------Literal_Array_[NULL, '10485.76', '25.5', NULL], 0x00006110009d0c20
------------Literal_NULL, 0x0000611000abaaa0
------------Literal_NULL, 0x0000611000440b20
----------Literal_'65537', 0x0000611000b9aa60
------Function_toDate, 0x00006110005eba58
-------ExpressionList, 0x0000607000017338
--------Literal_Array_[NULL, '0.0000000255', NULL], 0x00006110005a2360
--------Literal_'922337203685477580.7', 0x0000611000559660
--------Literal_NULL, 0x0000611000b80160
--------Literal_NULL, 0x0000611000822c20
'
Text-2 (AST-2 formatted):
'WITH NULL AS a SELECT toDate(-7 IN (inf - NULL, toDate([NULL, '10485.76', '25.5', NULL], NULL, NULL), '65537'), toDate([NULL, '0.0000000255', NULL], '922337203685477580.7', NULL, NULL))'
Text-1 must be equal to Text-2, but it is not.
```

### PR DESCRIPTION
**Pull request by akuzm**
Fixes https://github.com/ClickHouse/ClickHouse/issues/22116

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Parsers/ASTFunction.cpp [hunk 1]
```diff
@@ -214,27 +215,62 @@ void ASTFunction::formatImplWithoutAlias(const FormatSettings & settings, Format
...
+                /* A particularly stupid case. If we have a unary minus before
+                 * a literal that is a negative number "-(-1)" or "- -1", this
+                 * can not be formatted as `--1`, since this will be
+                 * interpreted as a comment. Instead, negate the literal
+                 * in place.
+                 */
+                if (literal && name == "negate")
+                {
+                    written = applyVisitor(
+                        [&settings](const auto & value) {
```
Comment (akuzm) [inline src/Parsers/ASTFunction.cpp:233]:
```suggestion
                        [&settings](const auto & value)
                        {
```

File: src/Parsers/ASTFunction.cpp [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #include <Parsers/ASTFunction.h>
 
+#include <Common/FieldVisitorsAccurateComparison.h>
```
Comment (akuzm) [inline src/Parsers/ASTFunction.cpp:3]:
```suggestion
```

File: src/Parsers/ASTFunction.cpp [hunk 3]
```diff
@@ -214,27 +214,74 @@ void ASTFunction::formatImplWithoutAlias(const FormatSettings & settings, Format
...
+                            }
+                            else if constexpr (std::is_arithmetic_v<ValueType>)
+                            {
+                                if (value >= 0)
+                                {
+                                    return false;
+                                }
+                                // We don't need parentheses around a single
+                                // literal.
+                                settings.ostr << -value;
```
Comment (qoega) [inline src/Parsers/ASTFunction.cpp:258]: UBSan complains about this negation

Comment (akuzm) [reply src/Parsers/ASTFunction.cpp:258]: Yes, I have seen the report.

**Comment by akuzm** [comment]
fuzzer https://github.com/ClickHouse/ClickHouse/issues/22290

**Pull request merged**